multik-jvm/src/main/kotlin/org/jetbrains/kotlinx/multik/jvm/pludecomposition.kt

```diff
@@ -0,0 +1,259 @@
+
+    if (a.shape[1] == 1) {
+        var imax = 0
+        var elemmax = a[0, 0]
+        for (i in 1 until a.shape[0] ) {
+            if (abs(a[i, 0]) > abs(elemmax)) {
+                elemmax = a[i, 0]
+                imax = i
+            }
+        }
+
+        if (elemmax != 0.0) {
+            // pivoting
+            a[0, 0] = a[imax, 0].also {
+                a[imax, 0] = a[0, 0]
+            }
+            rowPerm[0] = imax
+
+            for (i in 1 until a.shape[0] ) {
+                a[i, 0] /= elemmax
+            }
+        }
+
+        return
+    }
+
+    // apply recursively to [ a11 ]
+    //                      [ a21 ]
+    PLUdecomposition2Inplace(a[0..a.shape[0], 0..n1] as D2Array<Double>, rowPerm[0..min(a.shape[0], n1)] as D1Array<Int>)
+
+    // change [ a12 ]
+    //        [ a22 ]
+    for (i in rowPerm.indices) {
+        if (rowPerm[i] != 0) {
+            for (j in n1 until a.shape[1]) {
+                a[i, j] = a[i + rowPerm[i], j].also {
+                    a[i + rowPerm[i], j] = a[i, j]
+                }
+            }
+        }
+    }
+
+    // change a12
+    solveLowerTriangleSystemInplace(
+        a[0 .. n1, 0  .. n1     ] as D2Array<Double>,
+        a[0 .. n1, n1 .. n1 + n2] as D2Array<Double>
+    )
+
+    // update a22
+    dotThenPlusInplace(
+        a[n1 .. n1 + a.shape[0] - n1, 0  .. n1     ] as D2Array<Double>,
+        a[0  .. n1                  , n1 .. n1 + n2] as D2Array<Double>,
+        a[n1 .. n1 + a.shape[0] - n1, n1 .. n1 + n2] as D2Array<Double>,
+        -1.0,
+        1.0
+    )
+
+    // factor a22
+    PLUdecomposition2Inplace(a[n1 .. a.shape[0], n1 .. a.shape[1]] as D2Array<Double>,  rowPerm[n1 .. min(a.shape[0], a.shape[1])] as D1Array<Int>)
+
+    // apply rowPerm to a21
+    for (i in n1 until a.shape[0] ) {
+        if(i >= rowPerm.size) {
+            break
+        }
+        if (rowPerm[i] != 0) {
+            for (j in 0 until n1) {
+                a[i, j] = a[i + rowPerm[i], j].also {
+                    a[i + rowPerm[i], j] = a[i, j]
+                }
+            }
+        }
+    }
+
+}
+
+
+fun PLU(a: D2Array<Double>): Triple<D2Array<Double>, D2Array<Double>, D2Array<Double>> {
+    val _a = a.deepCopy()
+    val perm = mk.d1array<Int>(min(_a.shape[0], _a.shape[1])){ 0 }
+
+
+    PLUdecomposition2Inplace(_a, perm)
+    // since previous call _a contains answer
+
+    val L = mk.d2array(_a.shape[0], min(_a.shape[0], _a.shape[1])) { 0.0 }
+    val U = mk.d2array(min(_a.shape[0], _a.shape[1]), _a.shape[1]) {0.0}
+
+    for (i in 0 until L.shape[1]) {
+        L[i, i] = 1.0
+        for (j in 0 until i) {
+            L[i, j] = _a[i, j]
+        }
+    }
+    for (i in L.shape[1] until L.shape[0]) {
+        for (j in 0 until L.shape[1]) {
+            L[i, j] = _a[i, j]
+        }
+    }
+
+    for (i in 0 until U.shape[0]) {
+        for (j in i until U.shape[1]) {
+            U[i, j] = _a[i, j]
+        }
+    }
```

Comment:
map

Reply:
I don't know how to do it with map (please check second loop bounds)